client/js/lounge.js

```diff
@@ -1105,13 +1105,35 @@ $(function() {
 	chat.on("input", ".search", function() {
 		var value = $(this).val().toLowerCase();
 		var names = $(this).closest(".users").find(".names");
-		names.find(".user").each(function() {
-			var btn = $(this);
-			var name = btn.text().toLowerCase().replace(/[+%@~]/, "");
-			if (name.indexOf(value) > -1) {
-				btn.show();
+		// Find all the categories of users and loop through them.
+		names.find(".user-mode").each(function() {
+			// Declare array to store names to hide, in case the whole category can't be hidden.
+			var namesToHide = [];
+			var userModeDiv = $(this);
+			// Declare boolean variable to establish if the current category should be hidden.
+			var isDivToHide = true;
+			// Find and loop through each name.
+			userModeDiv.find(".user").each(function() {
+				var btn = $(this);
+				var name = btn.text().toLowerCase().replace(/[+%@~]/, "");
+				// Be sure to show matched names and relative category;
+				// store names to hide, in case it's not possible to hide the whole category.
+				if (name.indexOf(value) > -1) {
+					btn.show();
+					isDivToHide = false;
+				} else {
+					namesToHide.push(btn[0]);
+				}
+			});
+			// If possible, hide the whole category, otherwise hide the "names to hide".
+			if (isDivToHide) {
+				userModeDiv[0].style.display = "none";
 			} else {
-				btn.hide();
+				userModeDiv[0].style.display = "block";
+				var i;
+				for (i in namesToHide) {
```

Comment:
Right, I just found out that the for/in loop is slower and it is meant for objects, not arrays. My bad.